### DISCUSSION TITLE
Globs in extra paths

### DISCUSSION THREAD

**Discussion by itsdani** [microsoft/pylance-release#2712] [Enhancement]
I'm trying to setup a python monorepo build with bazel, and I'm struggling to make pylance recognize the library imports. I have managed to make it work by adding e.g. `<bazel-myproject>/external/<my-pip-dependencies>/pypi__numpy` to `python.analysis.extraPaths`, but adding a line for each dependency on every computer we use doesn't seem to be right.

My idea/request is that if we could add something like `dependencies/*/` or `dependencies/pypi__*/` to the extra paths, it would solve this problem entirely. The expected behavior would be to add every (matching) directory under `dependencies/` to the extra paths.

**Comment by judej** [comment]
Thanks for the request @itsdani, we will review the request and look at what it would take to get this functionality

**Comment by jakebailey** [comment]
Copying my info here from https://github.com/microsoft/pyright/issues/1684#issuecomment-806089804

> One consideration is that file watching logic for this becomes a little complicated because 1) we may need special handlers for parent directories of globbed items to know when the paths have to be regenerated, and 2) glob patterns may match files which we don't want in extraPaths (besides maybe zips if those ever get supported).

**Comment by dermidgen** [comment]
This is pretty critical for supporting monorepos with multiple python services.

Support for this feature would also resolve many of the open issues related to `poetry` with the use of local envs. 

My teams currently address this challenge with a bash script that is run whenever a new python service or library is added.

```bash
#!/bin/bash
shopt -s dotglob
shopt -s nullglob

VENV_PATHS_ARR=(*/*/.venv/lib/*/site-packages/)
VENV_PATHS=$(printf ",\"%s\"" "${VENV_PATHS_ARR[@]}")
EXTRA_PATHS=${VENV_PATHS:1}

jq ".\"python.analysis.extraPaths\" = [${EXTRA_PATHS}]" .vscode/settings.json > .vscode/settings.tmp
mv .vscode/settings.tmp .vscode/settings.json 
```

**Comment by jakebailey** [comment]
> Support for this feature would also resolve many of the open issues related to poetry with the use of local envs.

Can you clarify what you mean? I wouldn't expect anyone to use extraPaths to add paths that point to an interpreter.

**Comment by dermidgen** [comment]
We found `extraPaths` was the simplest way to support import resolution for small-medium monorepos.  For larger monorepos or where granular ACLs are required, we found multi-root workspaces to be useful.  However, even with multi-root workspaces, some roots may contain 10's to hundreds of services or libraries.

If there's a better approach than `extraPaths` - like actually switching envs - I would be very pleased to follow that. The downside to this approach is that imports will resolve regardless of which project your in.

With `poetry` and the following scaffolding:
```
./
...
./libs
./libs/lib-one
./libs/lib-one/poetry.toml
./libs/lib-one/pyproject.toml
./libs/lib-one/.venv
...
./stack
./stack/service-one
./stack/service-one/poetry.toml
./stack/service-one/pyproject.toml
./stack/service-one/.venv
...
```

Using `poetry.toml`
```
[virtualenvs]
in-project = true
```

**Comment by jakebailey** [comment]
You may be interested in https://github.com/microsoft/pyright/blob/main/docs/configuration.md#execution-environment-options, though I don't know to what extent it will have the behavior you're seeking. Execution environments are what @erictraut's team uses to manage their monorepo.

**Comment by dermidgen** [comment]
Execution environments are indeed helpful, in that it resolves the issue of resolving imports from outside the project path. This does require switching from `.vscode/settings.json` to `pyproject.toml` at the root of the repo - no biggie.

Unfortunately, since it doesn't allow setting `venv` for each execution environment :hand_over_mouth:, we still need to set `extraPaths` for each execution environment. Which, yes, is better than using `.vscode/settings.json`.  However, since we need to set `extraPaths` we'll still need a helper script to do a glob drill down for `"*/*/.venv/lib/*/site-packages/"`.

So I guess, if we can't get `executionEnvironments` to support an override for `venv`, we'd still greatly appreciate glob support for `extraPaths`.

**Comment by jakebailey** [comment]
The execution environment model is not suited to have separate interpreters, unfortunately. There used to be support, but it was broken as it was never intended to work. Multi-root runs separate analyzers, so they are able to have separate interpreters. Execution environments are within an analyzer, and can only have one.

We need to come up with a spec to better reconcile the idea of execution environments and multiroot into something that can capture how people are using. I think details on how your monorepo is laid out and what you're expecting would be helpful for that (though, this issue was about globbing, not that redesign 🙂)

**Comment by dermidgen** [comment]
@jakebailey I'd be happy to support that effort in any way; we can move the interpreters/monorepo discussion.  Here's an example of what I'm currently testing wtih. https://github.com/dermidgen/python-monorepo

I yield the thread back to globbing :grin:

**Comment by danmx** [comment]
any updates?

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by oselcuk-iqm** [comment]
Are there any updates on this issue? It's quite painful to use VSCode with a bazel python repo without this feature

**Reply by gdippolito** [reply]
I ended up creating a script to manipulate `.vscode/settings.json` for reference https://gist.github.com/gdippolito/72fc9e099d9c0d99ff0d79fac37864f9

We run the script after a bazel rule that will pull down all the dependencies.

**Comment by thomassajot** [comment]
It would be great to have this feature.

**Comment by guntribam** [comment]
Just keep the begging alive

**Comment by sulekz** [comment]
Additional begging.

**Comment by erictraut** [comment]
For those of you who are interested in this feature, it would be useful if you could provide some details about the problem you're trying to solve. That would allow us to consider other potential solutions.

There's a good reason why we don't support globs in extraPaths today. The order of extraPaths is significant. This setting defines the specific order in which imports are resolved. This ordering matters, for the same reason that the order of `sys.path` matters at runtime. Globs don't allow for any ordering. They are resolved in whatever order the file system happens to return directories or files. For that reason, it's not a good idea to use globs for any configuration settings that are order-dependent.

**Reply by misha-antonenko** [reply]
The glob-specified paths could be sorted, and the ordering would be neither too surprising nor inconsistent. It anyway sounds like a bad a idea to rely on globs when you want definite ordering

**Comment by itsdani** [comment]
My original problem was that when I set up a monorepo project with Bazel, it collects the dependencies of the projects into directories like `./dependencies/pip/<package-name>/`. So the executables are not in a single directory, and the list of directories can easily grow/change if someone adds a new dependency. In this scenario I would not care at all about the order of the extraPaths entries, because they are all different packages that should not conflict.

I suspect that extraPaths was never intended to do this kind of dependency inclusion. Virtual envs would be a better thing to do, but (when I tried doing this) Bazel did its own thing without using venvs.

**Reply by erictraut** [reply]
If bazel places all dependences into a directory named `./dependencies/pip/<package-name>`, can't you simply add `dependencies/pip` to your `extraPaths`?

Are you able to run your code in the environment that bazel creates? In other words, do imports resolve at runtime within this environment? If so, do you know how that works? For example, is bazel setting the `PYTHONPATH` environment variable to this directory (or directories) so the Python interpreter updates `sys.path` at startup?

**Comment by triplepoint** [comment]
For what it's worth, I have exactly the same use case: Bazel has an opinionated way of managing Python dependencies, and the only current solution is to regularly regenerate extraPaths with a shell script. Globbing would be the most obvious way to dynamically declare this list, but I'd be curious to hear alternative suggestions.  

Oh, and I'm one of the poor souls stuck using bazel in a polyrepo (not monorepo) scenario, so I think it's best to think of the bazel/globbing and the monorepo discussions as separate but related.

**Comment by rugged-robotics** [comment]
One use case where this is useful is for people using ROS1 devel workspaces. ROS1 uses a model of masking workspaces for installation and development, with the primary installation workspace manually installing files to a known location, and the development workspaces linked to those shadowing the installed files via some path magic and link to the package directories directly.

Since the development workspace contains multiple packages all at the same level (all code is in `workspace/src/` directory in discrete packages) as well as each ros package having it's own `src` directory which contains the python packages it defines; there's no real concern over resolution order for this application, but there is a need to effectively product a glob of `workspace/src/*/src/` or `workspace/src/**/src/` to have an extra path pointing inside each packages src directory.

It would be useful for this application is the glob in extra paths expanded in place in the list of extra paths, so we could prioritize python packages in the system and installation spaces, but all packages within the glob have arbitrary dependency order.

This may all seem somewhat convoluted, but it's the framework this system provides

**Comment by judej** [comment]
This has been added to our backlog. You can track progress at #5364

**Comment by adanielnoel** [comment]
Still an issue. Managed to get a workaround similar to [gdippolito](https://github.com/gdippolito)'s

> I ended up creating a script to manipulate `.vscode/settings.json` for reference https://gist.github.com/gdippolito/72fc9e099d9c0d99ff0d79fac37864f9
> 
> We run the script after a bazel rule that will pull down all the dependencies.

Instead, I created a py_binary rule that must depend on all external python packages I use like so:

```starlark
py_binary(
    name = "vscode_import_generator",
    srcs = ["vscode_import_generator.py"],
    deps = [
        "@pypi//numpy",
        ...
    ],
)
```

A bit annoying to have to keep it updated but couldn't find a better way since we can't glob dependencies in Bazel.

This is my modified version of the script: https://gist.github.com/gdippolito/72fc9e099d9c0d99ff0d79fac37864f9

**Comment by heejaechang** [comment]
This discussion seems to be divided into two requests: one for Mono users and the other for Bazel users. We probably need two separate solutions for each group. For Mono repo users, [this](https://github.com/microsoft/vscode-python/issues/21204#issuecomment-2096268938) will probably solve your issue?

**Comment by dududko** [comment]
Would like to post here a different workaround that allows to keep the `.vscode/settings.json` file small and instead generate the external repository that will contain symbolic links to external python packages.

https://gist.github.com/dududko/4cacb6172f069b0e439af69328bce907

You can find this solution in action in the [following](https://github.com/dududko/sortcpp/tree/298c7b24336c63c4b477de652a8251605328ccc7) git repository

**Reply by dududko** [reply]
If anyone use `rules_pycross` the `.vscode/settings.json` file should look like this
```
{
    "python.analysis.extraPaths": [
        "${workspaceFolder}/bazel-bin/external/rules_pycross++lock_file+poetry_lock_file/deps/numpy@2.2.6/site-packages",
        "${workspaceFolder}/bazel-bin/external/rules_pycross++lock_file+poetry_lock_file/deps/pandas@2.3.2/site-packages"
    ]
}
```

**Reply by dududko** [reply]
Try to solve a case when in my package I have a python code which is generated by build target based on jinja2 template and is imported into a static code. No matter what I do both Pyright and Pylance do not recognize the imports. If someone has a similar case please tell me how you made it work.

**Comment by mksetaro** [comment]
Hello!

Don't know if people are still looking at this but:

After struggling a bit to get vscode ty extension work in bazel monorepo (for symbols lookup in externals), I came up with a creative and gentle solution:

I've created a VSCode task running a shell script (thanks coding assistant) that goes through the external folder and looks for all site packages (it's a bit hacky but with bzlmod pip managed dependencies are stored using the pretty consistent pattern "rules_python++pip+") and interpreter, then it creates simlinks in a folder at root level which can be easily plugged into ty.toml.

This still requires to run a build in order to have bazel download deps, I've made a version using bazel query to selectively collecting pip packages but it's utterly slow. I find a good tradeoff that to lookup symbols I need to build the targets I'm working on.

Anyway, I'll leave below the snippets of my ty.toml, tasks.json and symlink script. I hope it can be helpful for you as well!

### tasks.json
```json
{
    "tasks": [
        {
            "label": "Python Linting: Symlink Pip Hubs",
            "type": "shell",
            "command": "./tools/symlink_pip_deps.sh",
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}
```

### symlink_pip_deps.sh
```bash
#!/bin/bash
set -e

DEST_DIR=".py_external_symlink"
EXTERNAL_ROOT=$(bazel info output_base)/external

echo "🔎 Scanning Bazel external for Python components..."

# Iterate through all rules_python managed repositories using the Bzlmod pattern
for REPO_PATH in "$EXTERNAL_ROOT"/rules_python[+][+]*; do
    [ -d "$REPO_PATH" ] || continue
    REPO_NAME=$(basename "$REPO_PATH")

    # We look for the python toolchain repositories (usually ending in _host or _target)
    if [[ "$REPO_NAME" == *"python_"*"_host"* ]] || [[ "$REPO_NAME" == *"python_"*"_target"* ]]; then
        PYTHON_BIN="$REPO_PATH/bin/python3"
        if [ -f "$PYTHON_BIN" ]; then
            echo "📝 Found Hermetic Interpreter: $REPO_NAME"
            echo "   └─ Mapping binary to $DEST_DIR/bin/python"
            
            ln -sf "$PYTHON_BIN" "$DEST_DIR/bin/python"
            ln -sf "$PYTHON_BIN" "$DEST_DIR/bin/python3"
        fi
    fi

    # Look for site-packages directories inside the pip hub or wheel repos
    SITE_PKG=$(find "$REPO_PATH" -maxdepth 4 -type d -name "site-packages" 2>/dev/null | head -n 1)
    if [ -n "$SITE_PKG" ]; then
        echo "📦 Found Pip Repository: $REPO_NAME"
        echo "   └─ Merging site-packages into $DEST_DIR"
        
        # cp -as creates the symlink forest (merges directories)
        cp -as "$SITE_PKG"/* "$DEST_DIR/" 2>/dev/null || true
    fi
done

echo -e "\n✅ IDE sync complete."
echo "   Interpreter path: ${PWD}/${DEST_DIR}/bin/python"
echo "   Package path:     ${PWD}/${DEST_DIR}"
```

### ty.toml
```toml
[environment]
python-version = "<your_python_version>"
root = ["."]
extra-paths = [".py_external_symlink"]

```

Cheerio!
